core-rust/state-manager/src/protocol/protocol_update.rs

```diff
@@ -0,0 +1,489 @@
+// This file contains the protocol update logic for specific protocol versions
+
+use node_common::locks::{LockFactory, RwLock, StateLock};
+use radix_engine::track::StateUpdates;
+use radix_engine::transaction::CostingParameters;
+use radix_engine_common::network::NetworkDefinition;
+use radix_engine_common::prelude::Decimal;
+use std::ops::Deref;
+use std::sync::Arc;
+
+use transaction::prelude::TransactionPayloadPreparable;
+use transaction::validation::{NotarizedTransactionValidator, ValidationConfig};
+use utils::btreemap;
+
+use crate::epoch_handling::EpochAwareAccuTreeFactory;
+use crate::mainnet_updates::MainnetProtocolUpdaterFactory;
+use crate::traits::{
+    CommitBundle, CommitStore, CommittedTransactionBundle, HashTreeUpdate, QueryableProofStore,
+    ReceiptAccuTreeSliceV1, SubstateStoreUpdate, TransactionAccuTreeSliceV1,
+};
+use crate::transaction::{
+    ExecutionConfigurator, FlashTransactionV1, LedgerTransaction, LedgerTransactionValidator,
+    PreparedLedgerTransaction, TransactionSeriesExecutor,
+};
+use crate::{
+    CommittedTransactionIdentifiers, ExecutionCache, LedgerHeader, LedgerProof, LedgerProofOrigin,
+    LoggingConfig, ProtocolState, StateManagerDatabase,
+};
+
+pub trait ProtocolUpdaterFactory {
+    fn supports_protocol_version(&self, protocol_version_name: &str) -> bool;
+
+    fn updater_for(
+        &self,
+        protocol_version_name: &str,
+        store: Arc<StateLock<StateManagerDatabase>>,
+    ) -> Box<dyn ProtocolUpdater>;
+}
+
+/// Protocol update consists of two events:
+/// 1) Updating the current (state computer) configuration ("transaction processing rules").
+///    This includes: transaction validation, execution configuration, etc
+/// 2) Executing arbitrary state updates against the current database state.
+///    While the abstraction is quite flexible, the only concrete implementation at the moment
+///    only modifies the state through committing system transactions (e.g. substate flash).
+pub trait ProtocolUpdater {
+    fn protocol_version_name(&self) -> String;
+    fn state_computer_configurator(&self) -> StateComputerConfigurator;
```

Comment:
This could also be an associated constant to make it clear that this shouldn't change.